### PR TITLE
Add `entries` and `from_entries`, similar to JavaScript's `Object.entries` and `Object.fromEntries`

### PR DESCRIPTION
**Pull request by someonewithpc**
Previous title: Add `enumerate`, similar to Python's

This PR adds an `enumerate(Transversable|array, int=)` function that acts like [Python's `enumerate`](https://docs.python.org/3.8/library/functions.html#enumerate)

```php
foreach (enumerate($this->hash) as $i => [$k => $v]) {
    // ...
}
```

My use case is to iterate over an array with string keys and have an associated integer that can be used as an ID

### CODE REVIEW

**Comment by lstrojny** [comment]
I like the idea! I would model it after JavaScripts `Object.fromEntries()` and `Object.entries()` function and have both functions available.

**Comment by someonewithpc** [comment]
@lstrojny I'm not too versed in JavaScript, but [from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#return_value):

> An array of the given object's own enumerable string-keyed property [key, value] pairs

So I don't see the connection. These may also be useful, for sure, but unless I'm not understanding it right, it'd be a completely different pair of functions

**Comment by lstrojny** [comment]
Just to clarify: `Object.entries()` does exactly what your (and Python's) `enumerate()` does and `Object.fromEntries()` does the opposite.

So this is how it could look for Functional:

```
Functional\entries(['foo' => 'bar', 'baz' => 'bla']); // [['foo', 'bar'], ['baz', 'bla']]
Functional\from_entries([['foo', 'bar'], ['baz', 'bla']]); // ['foo' => 'bar', 'baz' => 'bla']
```

**Comment by someonewithpc** [comment]
Ah, I see! Yes, that would be useful, but there's a subtle difference in the way I made it:

```php
Functional\enumerate(['H' => `'Hydrogen']);` // [0 => ['H' => 'Hydrogen]]
```

Difference being that it keeps the key-value map, meaning it can be reversed with `array_merge(...$enumerated)`.

I'm not entirely sure which is more useful in general. The one I implemented was the one I wanted, but I'll defer to you as to which you prefer/think is more useful

**Comment by jdreesen** [comment]
Not sure if this destructuring syntax is legal in PHP, though.

```php
foreach (enumerate($this->hash) as $i => [$k => $v]) {
    // ...
}
```

**Comment by someonewithpc** [comment]
@lstrojny I updated the code as per your recommendations 

The failing tests seem to be unrelated to my changes (I hope)

File: src/Functional/Entries.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * @link      https://github.com/lstrojny/functional-php
+ */
+
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * Inspired by JavaScript's `Object.entries`, and Python's `enumerate`,
```
Comment (lstrojny) [inline src/Functional/Entries.php:17]:
```suggestion
 * Inspired by JavaScript’s `Object.entries`, and Python’s `enumerate`,
```

File: src/Functional/FromEntries.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * @link      https://github.com/lstrojny/functional-php
+ */
+
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * Inspired by JavaScript's `Object.from_entries`,
```
Comment (lstrojny) [inline src/Functional/FromEntries.php:17]:
```suggestion
 * Inspired by JavaScript’s `Object.fromEntries`,
```

File: src/Functional/FromEntries.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * @param Traversable|array $collection
+ * @return array
+ * @no-named-arguments
+ */
+function from_entries($collection)
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $aggregation = [];
+    foreach ($collection as [$key, $value]) {
```
Comment (lstrojny) [inline src/Functional/FromEntries.php:30]: Let's throw an exception if it's not a pair

File: src/Functional/Entries.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * convert a key-value map into an array of key-value pairs
+ *
+ * @see Functional\from_entries
+ * @param Traversable|array $collection
+ * @param int $start
+ * @return array
+ * @no-named-arguments
+ */
+function entries($collection, int $start = 0)
```
[the remark this answers is not in the document]
Comment (someonewithpc) [reply src/Functional/Entries.php:26]: `enumerate` supports it, and I was going to use it to have the list start at the ID of the next database row

File: src/Functional/Entries.php (hunk 1 above)
Comment (someonewithpc) [reply src/Functional/Entries.php:17]: Is that U+2019: RIGHT SINGLE QUOTATION MARK? Why that change?

File: src/Functional/FromEntries.php (hunk 3 above)
Comment (someonewithpc) [reply src/Functional/FromEntries.php:30]: You mean if collection doesn't have string keys?

File: src/Functional/Entries.php (hunk 1 above)
Comment (jchase) [reply src/Functional/Entries.php:17]: It's typographically correct.

File: src/Functional/FromEntries.php (hunk 3 above)
Comment (someonewithpc) [reply src/Functional/FromEntries.php:30]: Right, reading this again I see what you mean. Makes sense, but I'm not sure how to check that

Comment (lstrojny) [reply src/Functional/FromEntries.php:30]: Something like this:
```
foreach ($collection as $enty) {
     InvalidArgumentException::assertPair($entry);
}
```

```
public static function assertPair($pair): void
{
    if (!(is_array($pair) || $pair instanceof ArrayAccess) && !isset($pair[0], $pair[1])) {
          … throw
    }
}
```

Then add a few test cases for the `assertPair()` method

**Comment by someonewithpc** [comment]
I've made the changes you requested, I believe. Apologies for the delay

File: src/Functional/FromEntries.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+function from_entries($collection)
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $aggregation = [];
+    foreach ($collection as $entry) {
+        InvalidArgumentException::assertPair($entry, __FUNCTION__, 1);
+        [$key, $value] = $entry;
+        $aggregation[$key] = $value;
```
Comment (lstrojny) [inline src/Functional/FromEntries.php:33]: One more thing: we should also check that `$key` is fine to be used as an array key, so checking that it's a string or an int

File: src/Functional/FromEntries.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * Inspired by JavaScript’s `Object.fromEntries`,
+ * convert an array of key-value tuples into a key-value map
```
Comment (lstrojny) [inline src/Functional/FromEntries.php:18]:
```suggestion
 * convert an array of key-value pairs into a key-value map
```

File: src/Functional/Entries.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (lstrojny) [reply src/Functional/Entries.php:26]: I am in favor of removing it, if somebody needs to reindex they can do that in a separate call

**Review by lstrojny — CHANGES_REQUESTED** [review]
Looking very good already. Few small things, see inline. Also: could you update the docs to include it?

**Comment by someonewithpc** [comment]
I believe it's finally done :)

**Comment by lstrojny** [comment]
Great, thank you for your patience!

**Comment by lol768** [comment]
@lstrojny Has this been released?

**Pull request merged**